### PR TITLE
Implement the restriction on calls to non-escaping function parameter…

### PR DESCRIPTION
**Pull request by rjmccall**
…s from the SE-0176 revision.

This is the swift-4.0-branch version of #10391.  It fixes rdar://32118950.

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@jrose-apple Mind reviewing this?

File: lib/Sema/MiscDiagnostics.cpp [hunk 1]
```diff
@@ -446,9 +450,118 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
...
+        }
+
+        if (auto fn = dyn_cast<AbstractFunctionDecl>(decl)) {
+          if (fn->getDeclContext()->isLocalContext()) {
+            return findInCaptures(tc, fn);
+          }
+          return {};
+        }
+
+        // FIXME: captures of computed local vars?  Can these be non-escaping?
```
Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: Currently they can't be, but they probably should be…

```swift
func test(x: () -> Int) {
  var y: Int { return x() + 1 }
  print(y)
}
test(x: { 1 })
```
> error: declaration closing over non-escaping parameter 'x' may allow it to escape

**Review by jrose-apple — COMMENTED** [review]
Seems reasonable to me. What is the "Dispatch problem" that we still need to fix?

**Comment by rjmccall** [comment]
Dispatch violates the rule.  I've already asked them to take a look.

https://github.com/apple/swift-corelibs-libdispatch/blob/master/src/swift/Queue.swift#L219

File: lib/Sema/MiscDiagnostics.cpp (hunk 1 above)
Comment (rjmccall) [inline lib/Sema/MiscDiagnostics.cpp]: How do we record captures here?  Just capture lists on the various accessors?

Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: Seems to be, from an AST dump.

Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: Note that captures on a `didSet` are *not* propagated up to the setter.

```swift
func test(x: @escaping () -> Int) {
  var y: Int = 0 {
    didSet { print(x() + 1) }
  }
  y = 2
}
test(x: { 1 })
```

```
(var_decl "y" type='Int' interface type='Int' access=private storage_kind=stored_with_observers
  (func_decl implicit 'anonname=0x7ffec2062c90' interface type='() -> Int' access=private captures=(y<direct>) getter_for=y
    [synthesized body])
  (func_decl implicit 'anonname=0x7ffec2062e90' interface type='(Int) -> ()' access=private captures=(y<direct>, _) setter_for=y
    (parameter_list
      (parameter "newValue" type='Int' interface type='Int'))
    (brace_stmt
      (pattern_binding_decl implicit
        (pattern_named implicit type='Int' 'tmp')
        (load_expr implicit type='Int'
          (declref_expr implicit type='@lvalue Int' accessKind=read decl=main.(file).test(x:).y@<stdin>:2:7 direct_to_storage function_ref=unapplied)))

      (var_decl implicit "tmp" type='Int' interface type='Int' access=private let storage_kind=stored)

      (assign_expr
        (declref_expr implicit type='@lvalue Int' accessKind=write decl=main.(file).test(x:).y@<stdin>:2:7 direct_to_storage function_ref=unapplied)
        (declref_expr implicit type='Int' decl=main.(file).test(x:).<anonymous>.newValue@<stdin>:2:20 function_ref=unapplied))
      (call_expr implicit type='()' nothrow arg_labels=_:
        (declref_expr implicit type='(Int) -> ()' decl=main.(file).test(x:)._@<stdin>:2:20 function_ref=single)
        (paren_expr implicit type='(Int)'
          (declref_expr implicit type='Int' decl=main.(file).test(x:).<anonymous>.tmp function_ref=unapplied)))))
  (func_decl 'anonname=0x7ffec2062920' interface type='(Int) -> ()' access=private captures=(x) didSet_for=y
    [type-checked body]))
```

**Comment by rjmccall** [comment]
Description: This is an important piece of the static checking for SE-0176 that we hadn't implemented yet.  This patch implements the check as a warning in order to stage in the diagnostic and discover places that violate the new rule, like the one in Dispatch described above.

This work is being tracked by rdar://32118950.

This patch does not alter code-generation in any way; it is a compile-time check only, performed using an AST-level analysis.

The risk is low.  The analysis only emits a warning.  Furthermore, it is gated so that it does nothing unless there is a direct call to a non-escaping function parameter; therefore even if there is a logic error in the analysis which could e.g. lead to a compiler crash, it will only kick in for a code structure which is... I wouldn't go so far as to say "uncommon", but "less than ubiquitous" seems fair.

It has been tested using our regression tests.

**Comment by tkremenek** [comment]
@swift-ci test source compatibility

**Pull request merged**